File: py/pipedrive_sdk/features.py
# Pipedrive SDK feature factory

from pipedrive_sdk.feature.base_feature import PipedriveBaseFeature
from pipedrive_sdk.feature.debug_feature import PipedriveDebugFeature
from pipedrive_sdk.feature.idempotency_feature import PipedriveIdempotencyFeature
from pipedrive_sdk.feature.metrics_feature import PipedriveMetricsFeature
from pipedrive_sdk.feature.paging_feature import PipedrivePagingFeature
from pipedrive_sdk.feature.ratelimit_feature import PipedriveRatelimitFeature
from pipedrive_sdk.feature.retry_feature import PipedriveRetryFeature
from pipedrive_sdk.feature.test_feature import PipedriveTestFeature
from pipedrive_sdk.feature.timeout_feature import PipedriveTimeoutFeature


_FEATURES = {
    "base": lambda: PipedriveBaseFeature(),
    "debug": lambda: PipedriveDebugFeature(),
    "idempotency": lambda: PipedriveIdempotencyFeature(),
    "metrics": lambda: PipedriveMetricsFeature(),
    "paging": lambda: PipedrivePagingFeature(),
    "ratelimit": lambda: PipedriveRatelimitFeature(),
    "retry": lambda: PipedriveRetryFeature(),
    "test": lambda: PipedriveTestFeature(),
    "timeout": lambda: PipedriveTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES

File: php/features.php
<?php
declare(strict_types=1);

// Pipedrive SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class PipedriveFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new PipedriveBaseFeature();
            case "debug":
                return new PipedriveDebugFeature();
            case "idempotency":
                return new PipedriveIdempotencyFeature();
            case "metrics":
                return new PipedriveMetricsFeature();
            case "paging":
                return new PipedrivePagingFeature();
            case "ratelimit":
                return new PipedriveRatelimitFeature();
            case "retry":
                return new PipedriveRetryFeature();
            case "test":
                return new PipedriveTestFeature();
            case "timeout":
                return new PipedriveTimeoutFeature();
            default:
                return new PipedriveBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}
